File: myblogsfrontend/src/app/service/post.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from 'src/environments/environment';
import { PostSearchResponse } from '../payload/PostSearchResponse';
import { CategorySidebar } from '../payload/CategorySidebar';
import { DropdownResponse } from '../payload/DropdownResponse';
import { CustomHttpResponse } from '../payload/CustomHttpResponse';
import { PostCreateRequest } from '../payload/PostCreateRequest';
import { PostEditRequest } from '../payload/PostEditRequest';
import { PostEditViewResponse } from '../payload/PostEditViewResponse';

@Injectable({
  providedIn: 'root'
})
export class PostService {

  // 'http://localhost:8080'
  host = environment.apiUrl;

  constructor(
    private http: HttpClient
  ) { }


  // get list of posts by category id
  searchPosts(pageNumber: number, pageSize: number, categoryid: number): Observable<PostSearchResponse[]> {

    return this.http.get<PostSearchResponse[]>(
      `${this.host}/post-search?pageNumber=${pageNumber}&pageSize=${pageSize}&categoryid=${categoryid}`
    );

  } // end of searchPosts()

  // get all active categories
  getAllActiveCategories(): Observable<DropdownResponse[]> {

    return this.http.get<DropdownResponse[]>(
      `${this.host}/category-active`
    )

  } // end of getAllActiveCategories()

  // create a new post
  public createPost(postCreateRequest: PostCreateRequest): Observable<CustomHttpResponse> {

    return this.http.post<CustomHttpResponse>(`${this.host}/post-create`, postCreateRequest);

  } // end of createPost()

  // edit an existing post
  public editPost(postEditRequest: PostEditRequest): Observable<CustomHttpResponse> {

    return this.http.put<CustomHttpResponse>(`${this.host}/post-edit`, postEditRequest);

  } // end of editPost()

  // find post by id
  findById(id: number): Observable<PostEditViewResponse> {

    return this.http.get<PostEditViewResponse>(`${this.host}/post-list/${id}`);

  } // end of findById()

  // get number of posts of each category
  getCategoriesSidebar(): Observable<CategorySidebar[]> {

    return this.http.get<CategorySidebar[]>(
      `${this.host}/category-sidebar`
    );

  } // end of getCategoriesSidebar()

} // end of class PostService

File: myblogsbackend/src/main/java/com/ez/myblogsbackend/service/PostService.java
package com.ez.myblogsbackend.service;

import com.ez.myblogsbackend.entity.Post;
import com.ez.myblogsbackend.exception.BadDataException;
import com.ez.myblogsbackend.payload.*;
import com.ez.myblogsbackend.repository.PostRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import javax.persistence.EntityNotFoundException;
import java.util.List;

import static com.ez.myblogsbackend.constant.Constant.NO_POST_FOUND_BY_ID;
import static org.springframework.http.HttpStatus.OK;

@Service
public class PostService {
    private Logger LOGGER = LoggerFactory.getLogger(getClass());

    @Autowired
    private PostRepository postRepository;

    //
    // sidebar menu.
    //
    // get category sidebar(categories and their number of posts)
    public List<CategorySidebar> getCategorySidebar() {

        LOGGER.info("Get category sidebar)");

        return postRepository.getCategorySidebar();

    } // end of getCategorySidebar()

    // get all active categories
    public List<DropdownResponse> getAllActiveCategories() {

        LOGGER.info("Get all active categories)");

        return postRepository.getAllActiveCategories();

    } // end of getAllActiveCategories()

    // search posts by category id.
    public List<PostSearchResponse> searchPosts(long pageNumber, long pageSize, long categoryid) {

        LOGGER.info("Get posts based on category id");

        return postRepository.searchPosts(pageNumber, pageSize, categoryid);

    } // end of searchPosts()

    // create a new post.
    public HttpResponse createPost(PostCreateRequest postCreateRequest) {

        LOGGER.info("create a new post");
        LOGGER.info("Post is sent from client: " + postCreateRequest.toString());

        Post post = new Post();
        post.setCreatorid(postCreateRequest.getCreatorid());
        post.setTitle(postCreateRequest.getTitle());
        post.setCategoryid(postCreateRequest.getCategoryid());
        post.setContent(postCreateRequest.getContent());

        // 'Publish' status
        post.setPostStatusid(1);

        // save new post into the "post" table in database.
        postRepository.save(post);

        return new HttpResponse(OK.value(), "Post is created successful!");

    }

    // get post by post id.
    public PostEditViewResponse getPostById(Long id) {

        return postRepository.getPostById(id);

    } // end of getPostById()

    // update existing post.
    public HttpResponse updatePost(PostEditRequest postEditRequest)
            throws EntityNotFoundException, BadDataException {

        LOGGER.info("Update post");
        LOGGER.info("Post is sent from client: " + postEditRequest.toString());
        LOGGER.info("Post id: " + postEditRequest.getPostid());

        // get existing post
        Post existingPost = postRepository.findById(postEditRequest.getPostid())
                .orElseThrow(() -> new EntityNotFoundException(NO_POST_FOUND_BY_ID + postEditRequest.getPostid()));

        LOGGER.info("Existing post:" + existingPost.toString());

        // set new values to existing post
        existingPost.setTitle(postEditRequest.getTitle());
        existingPost.setCategoryid(postEditRequest.getCategoryid());
        existingPost.setContent(postEditRequest.getContent());

        postRepository.save(existingPost);

        return new HttpResponse(OK.value(),
                "Post '" + postEditRequest.getPostid() + "' was updated successful.");

    } // end of updatePost()

    // delete a post by its id
    public HttpResponse deletePost(Long id)
            throws EntityNotFoundException {

        // delete post by its id
        postRepository.deleteById(id);

        return new HttpResponse(OK.value(),
                "Post with id = " + id + "' was deleted successful.");

    } // end of deletePost()

}
